### DISCUSSION TITLE
SDL2-Mixer not linking correctly for Linux

### DISCUSSION THREAD

**Discussion by obscurelyme** [microsoft/vcpkg#21182] [Q&A]
Hello, 

I'm really struggling to find a way to include SDL2-Mixer with my project located [here](https://github.com/obscurelyme/CoffeeMaker/tree/feature/sdl2-mixer). Right now that branch compiles and works fine on Windows 10, but on Linux I'm seeing the following error...

```
[build] /bin/ld: vcpkg_installed/x64-linux/debug/lib/libSDL2_mixer.a(music_mpg123.c.o): in function `MPG123_Load':
[build] /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:87: undefined reference to `mpg123_close'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:88: undefined reference to `mpg123_delete'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:89: undefined reference to `mpg123_exit'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:90: undefined reference to `mpg123_format'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:91: undefined reference to `mpg123_format_none'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:92: undefined reference to `mpg123_getformat'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:93: undefined reference to `mpg123_init'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:94: undefined reference to `mpg123_new'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:95: undefined reference to `mpg123_open_handle'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:96: undefined reference to `mpg123_plain_strerror'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:97: undefined reference to `mpg123_rates'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:98: undefined reference to `mpg123_read'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:99: undefined reference to `mpg123_replace_reader_handle'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:100: undefined reference to `mpg123_seek'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_mpg123.c:101: undefined reference to `mpg123_strerror'
[build] /bin/ld: vcpkg_installed/x64-linux/debug/lib/libSDL2_mixer.a(music_ogg.c.o): in function `OGG_Load':
[build] /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_ogg.c:91: undefined reference to `ov_clear'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_ogg.c:92: undefined reference to `ov_info'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_ogg.c:93: undefined reference to `ov_comment'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_ogg.c:94: undefined reference to `ov_open_callbacks'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_ogg.c:95: undefined reference to `ov_pcm_total'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_ogg.c:100: undefined reference to `ov_read'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_ogg.c:101: undefined reference to `ov_time_seek'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_ogg.c:103: undefined reference to `ov_pcm_seek'
[build] /bin/ld: /home/obcurelyme/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396.clean/music_ogg.c:104: undefined reference to `ov_pcm_tell'
```
If I include `dynamic-load` the compilation error goes away but there is no .so file that gets copy/pasted into the `build` directory. I figured that this could be due to the fact that x64-linux tripet means static by default. But when I manually install SDL2-Mixer with the following command after following the instructions [here](https://github.com/microsoft/vcpkg/blob/master/docs/examples/overlay-triplets-linux-dynamic.md) my project won't configure because Cmake won't find SDL2-Mixer library. I don't know if this is due to the fact that I have Manifest mode turned on? SDL2-Mixer seems to be an odd one off library that's proving very difficult to work with cross platforms right now, and I'm scratching my head on how to get this built properly. Any guidance here would be greatly appreciated.

To recap, some of the things I've tried...

- blown away vcpkg entirely and repulled and rebuilt everything again.
- manually install SDL2-Mixer with `vcpkg install sdl2-mixer[mpg123]:x64-linux-dynamic` (project can no longer find SDL2-Mixer)

**Comment by obscurelyme** [comment]
Update:

Here is a snippet from my [CMakeLists.txt](https://github.com/obscurelyme/CoffeeMaker/blob/feature/sdl2-mixer/CMakeLists.txt#L98-L117) file 

```
elseif(LINUX_PLATFORM)
  message("Linking dependencies for Linux")
  add_compile_definitions(_FILE_OFFSET_BITS=64)
  find_package(Ogg CONFIG REQUIRED)
  find_package(Vorbis CONFIG REQUIRED)
  find_package(mpg123 CONFIG REQUIRED)

  target_link_libraries(main PRIVATE Ogg::ogg)
  target_link_libraries(main PRIVATE Vorbis::vorbis Vorbis::vorbisenc Vorbis::vorbisfile)
  target_link_libraries(main PRIVATE MPG123::libmpg123 MPG123::libout123 MPG123::libsyn123)

  # SDL2
  target_link_libraries(main PRIVATE SDL2::SDL2 SDL2::SDL2main SDL2::SDL2-static)
  # SDL2-image
  target_link_libraries(main PRIVATE SDL2::SDL2_image)
  # SDL2-ttf
  target_link_libraries(main PRIVATE SDL2::SDL2_ttf)
  # SDL2-mixer
  target_link_libraries(main PRIVATE SDL2::SDL2_mixer)
endif()
```

Unless I'm super tired, this is exactly what vcpkg specifies to put when installing the dependencies I have.

**Reply by dg0yt** [reply]
`SDL2::SDL2_mixer` seems to not declare required dependencies (upstream bug, port bug).
As a workaround, you might try to move the SDL2 lines before the Ogg/Vorbis/MPG123 lines. Static libraries need to be added in the right order.

**Reply by obscurelyme** [reply]
That did the trick! Jesus H. I knew this was something really small but just couldn't put my finger on it.
